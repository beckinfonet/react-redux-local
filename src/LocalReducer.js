import { Component } from 'react'
import { func, objectOf, arrayOf, object } from 'prop-types'
import { createStore, applyMiddleware, bindActionCreators } from 'redux'
import createSagaMiddleware from 'redux-saga'
import { composeWithDevTools } from 'redux-devtools-extension'

class LocalReducer extends Component {
  constructor (props) {
    super(props)

    const { reducer, sagas, middleware, actions, devToolsOptions } = props

    this.sagaMiddleware = sagas.length && createSagaMiddleware()

    const allMiddleware = [...middleware]

    if (this.sagaMiddleware) allMiddleware.push(this.sagaMiddleware)

    const enhancers = applyMiddleware(...allMiddleware)

    const composeEnhancers = composeWithDevTools(devToolsOptions)

    this.state = reducer(undefined, {})

    this.sagas = []

    this.store = createStore(reducer, devToolsOptions ? composeEnhancers(enhancers) : enhancers)
    this.store.subscribe(() => this.setState(this.store.getState()))

    this.dispatch = this.dispatch.bind(this)
    this.boundActions = bindActionCreators(actions, this.dispatch)
  }

  componentWillMount () {
    this.props.sagas.map(saga => this.sagas.push(this.sagaMiddleware.run(saga)))
  }

  componentWillUnMount () {
    this.sagas.map(saga => saga.cancel())
  }

  dispatch (action = {}) {
    this.store.dispatch(action)
  }

  render () {
    return this.props.children(this.state, this.boundActions)
  }
}

LocalReducer.propTypes = {
  reducer: func.isRequired,
  actions: objectOf(func.isRequired).isRequired,
  sagas: arrayOf(func.isRequired),
  middleware: arrayOf(func.isRequired),
  render: func.isRequired,
  devToolsOptions: object
}

LocalReducer.defaultProps = {
  sagas: [],
  middleware: []
}

export default LocalReducer
