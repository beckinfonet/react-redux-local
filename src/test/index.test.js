import React from 'react'
import { mount, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { createSerializer } from 'enzyme-to-json'
import { takeEvery } from 'redux-saga'

import LocalReducer from '../'
import { reducer, actions, types } from './duck'

configure({ adapter: new Adapter() })
expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }))

it('should be defined', () => {
  expect(LocalReducer).toBeDefined()
})

it('should render different states based on actions', () => {
  const wrapper = mount(
    <LocalReducer
      reducer={reducer}
      actions={actions}
      devToolsOptions={{ name: 'testing react-redux-local' }}
      render={({ counter }) => <h1>Counter: {counter}</h1>}
    />
  )

  /* -------------------> Increase <------------------- */
  expect(wrapper.instance().state).toMatchSnapshot()
  wrapper.instance().dispatch(actions.increase())
  expect(wrapper.instance().state).toMatchSnapshot()
  wrapper.instance().boundActions.increase()
  expect(wrapper.instance().state).toMatchSnapshot()

  /* -------------------> Reset <------------------- */
  wrapper.instance().dispatch(actions.reset())
  expect(wrapper.instance().state).toMatchSnapshot()

  /* -------------------> Decrease <------------------- */
  wrapper.instance().dispatch(actions.decrease())
  expect(wrapper.instance().state).toMatchSnapshot()
  wrapper.instance().boundActions.decrease()
  expect(wrapper.instance().state).toMatchSnapshot()

  /* -------------------> Reset <------------------- */
  wrapper.instance().boundActions.reset()
  expect(wrapper.instance().state).toMatchSnapshot()
})

it('should call saga', () => {
  const mockCallback = jest.fn()

  function* saga() {
    yield takeEvery(types.INCREASE, mockCallback)
    yield takeEvery(types.DECREASE, mockCallback)
    yield takeEvery(types.RESET, mockCallback)
  }

  const wrapper = mount(
    <LocalReducer
      reducer={reducer}
      actions={actions}
      saga={saga}
      render={({ counter }) => <h1>Counter: {counter}</h1>}
    />
  )

  wrapper.instance().boundActions.increase()
  expect(mockCallback.mock.calls.length).toBe(1)

  wrapper.instance().boundActions.decrease()
  expect(mockCallback.mock.calls.length).toBe(2)

  wrapper.instance().boundActions.reset()
  expect(mockCallback.mock.calls.length).toBe(3)
})

it('middleware should work', () => {
  const mockCallback = jest.fn()

  const middleware = store => next => action => {
    mockCallback()
    return next(action)
  }

  const wrapper = mount(
    <LocalReducer
      reducer={reducer}
      actions={actions}
      middleware={[middleware]}
      render={({ counter }) => <h1>Counter: {counter}</h1>}
    />
  )

  wrapper.instance().dispatch({ type: 'TOTALY_FAKE_ACTION_1' })
  expect(mockCallback.mock.calls.length).toBe(1)
  wrapper.instance().dispatch({ type: 'TOTALY_FAKE_ACTION_2' })
  expect(mockCallback.mock.calls.length).toBe(2)
})
