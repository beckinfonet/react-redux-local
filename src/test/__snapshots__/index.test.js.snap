// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render different states based on actions 1`] = `
Object {
  "counter": 0,
}
`;

exports[`should render different states based on actions 2`] = `
Object {
  "counter": 1,
}
`;

exports[`should render different states based on actions 3`] = `
Object {
  "counter": 2,
}
`;

exports[`should render different states based on actions 4`] = `
Object {
  "counter": 0,
}
`;

exports[`should render different states based on actions 5`] = `
Object {
  "counter": -1,
}
`;

exports[`should render different states based on actions 6`] = `
Object {
  "counter": -2,
}
`;

exports[`should render different states based on actions 7`] = `
Object {
  "counter": 0,
}
`;
